local-tests/chainrunner/chain.py

```diff
@@ -0,0 +1,108 @@
+        cmd += flags_from_dict(kwargs)
+
+        chainspec = op.join(self.path, 'chainspec.json')
+        with open(chainspec, 'w') as f:
+            subprocess.run(cmd, stdout=f).check_returncode()
+
+        self.nodes = []
+        for a in accounts:
+            n = node(binary, chainspec, op.join(self.path, a), self.path)
+            n.flags['node-key-file'] = op.join(self.path, a, 'p2p_secret')
+            self.nodes.append(n)
+
+
+    def set_flags(self, *args, **kwargs):
+        """Set common flags for all nodes. Positional arguments are used as binary flags and should be strings.
+        Keyword arguments are translated to valued flags: `my_arg=some_val` results in `--my-arg some_val`
+        in the binary call. Seq (type alias for int) can be used to specify numerical values that should
+        be different for each node. `val=Seq(13)` results in `--val 13` for node0, `--val 14` for node1 and so on."""
+        for k in args:
+            for n in self.nodes:
+                n.flags[k] = True
+        for k,v in kwargs.items():
+            for i,n in enumerate(self.nodes):
+                n.flags[k] = v+i if isinstance(v, Seq) else v
+
+
+    def set_binary(self, binary, nodes=None):
+        """Replace nodes' binary with `binary`. Optional `nodes` argument can be used to specify
+        which nodes are affected and should be a list of integer indices (0..N-1). Affects all nodes if ommitted."""
+        if not op.isfile(binary):
+            raise FileNotFoundError(f'file not found: {binary}')
+        idx = nodes or range(len(self.nodes))
+        for i in idx:
+            self.nodes[i].binary = binary
+
+
+    def set_chainspec(self, chainspec, nodes=None):
+        """Replace nodes' chainspec with `chainspec`. Optional `nodes` argument can be used to specify
+        which nodes are affected and should be a list of integer indices (0..N-1). Affects all nodes if ommitted."""
+        if not op.isfile(chainspec):
+            raise FileNotFoundError(f'file not found: {chainspec}')
+        idx = nodes or range(len(self.nodes))
+        for i in idx:
+            self.nodes[i].chainspec = chainspec
+
+
+    def start(self, name, nodes=None):
+        """Start the chain. `name` will be used to name logfiles: name0.log, name1.log etc.
+        Optional `nodes` argument can be used to specify which nodes are affected and should be
+        a list of integer indices (0..N-1). Affects all nodes if ommitted."""
+        idx = nodes or range(len(self.nodes))
+        for i in idx:
+            self.nodes[i].start(name+str(i))
+
+
+    def stop(self, nodes=None):
+        """Stop the chain. Optional `nodes` argument can be used to specify which nodes are affected
+        and should be a list of integer indices (0..N-1). Affects all nodes if ommitted."""
+        idx = nodes or range(len(self.nodes))
+        for i in idx:
+            self.nodes[i].stop()
```

Comment:
optional, but may be fun: here you have 6 methods that are almost identical, so I would suggest something like this:
```python
def _apply(self, action, nodes):
  for node_id in (nodes or range(len(self.nodes))):
    action(self.nodes[node_id])

def stop(self, nodes=None):
  self._apply(lambda node: node.stop(), nodes)
```

Reply:
I thought about something similar, but going a bit further - to have `Chain` class read the `Node` deinition and automatically create, for each `Node` method, such a shortcut with additional `nodes` parameter. Needless to say, this is waaaaaaayyyy overcomplicated, so decided to just go with some code duplication. But I like your idea a lot and will implement it :)

Reply:
Nope, that would be too beautiful. You cannot reduce `start` to that (needs a different argument for each node) and `set_chainspec` and `set_binary` are not easily expressible as lambda expressions (you cannot do `lambda x: x.attribute = value`)

Reply:
You can have them in the closure/local functions, the action also may benefit from additional some `i` arg or alternatively the node should have some id inside it:
```py
    def set_binary(self, binary, nodes=None):
        ...
        def _set(node):
            node.binary = binary
        apply(lambda node, _node_id: _set(node), nodes)
```

Reply:
Yeah, but then we "simplify" a 3 line code-duplication by replacing it with more code AND an additional opaque step in between. This code is meant to be used as an utility tool, it's better to keep it simple and straightforward in case someone working with it wants to peek into source to quickly check something.

Reply:
fair, resolving